### PR TITLE
Fix fe dead lock

### LINKED ISSUE #6423: The process of FE hang while create resource.
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by DorianZheng**
Signed-off-by: dorianzheng <xingzhengde72@gmail.com>

## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/6423

## Problem Summary(Required) ：

The concurrent call of `HiveReposiroty::getClient` and `ResourceMgr::dropResource` might cause dead lock. Say there are two threads call above two method concurrently, the execution sequence is as below:

```
thread 1:                                            thread 2:
HiveRepository::getClient                            ResourceMgr::dropResource
                  |                 
  hiveRepository::metaClientsLock                      Resource::writeLock
                  |                                         |
  Resource::getResource                                onDropResource
                  |                                         |
                  |                                     HiveRepository::clearCache                                                            
     Resource::readLock                                     |
                                                         HiveRepository::metaClientsLock
``` 
the thread 1 will hold `HiveRepository::metaClientsLock` and wait on `Resource::readLock` while thread 2 will hold ` Resource::writeLock` and wait on `HiveRepository::metaClientsLock`. So we shouldn't call `HiveRepository::clearCache` inside the `Resource::writeLock`

### CODE REVIEW

**Comment by gengjun-git** [comment]
Why there is no lock hold in replayCreateResource? Seems it is a bug.

**Comment by DorianZheng** [comment]
> Why there is no lock hold in replayCreateResource? Seems it is a bug.

@gengjun-git Is `replayCreateResource` will be called concurrently with another method?

**Comment by gengjun-git** [comment]
> > Why there is no lock hold in replayCreateResource? Seems it is a bug.
> 
> @gengjun-git Is `replayCreateResource` will be called concurrently with another method?

The nameToResource will be concurrently called. When you are calling the `replayCreateResource`, there maybe a query calling the nameToResource.

**Comment by DorianZheng** [comment]
run starrocks_fe_unittest

**Review by gengjun-git — DISMISSED** [review]
LGTM

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ResourceMgr.java|7|7|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/catalog/ResourceMgr.java [hunk 1]
```diff
@@ -191,11 +202,17 @@ public void alterResource(AlterResourceStmt stmt) throws DdlException {
...
+            droppedResource = resource;
+
             GlobalStateMgr.getCurrentState().getEditLog().logCreateResource(resource);
         } finally {
             this.writeUnLock();
         }
+
+        // Do not invoke HiveRepository::clearCache inside `Resource.rwLock`. Otherwise, it might cause deadlock.
+        // Because HiveRepository::getClient will hold `Resource.rwLock` inside `HiveRepository::xxxLock`
+        GlobalStateMgr.getCurrentState().getHiveRepository().clearCache(droppedResource.getName());
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ResourceMgr.java:215]: Resource is not dropped actually.
Better to use name in AlterResourceStmt.

**Review by wyb — APPROVED** [review]

**Comment by DorianZheng** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/ResourceMgr.java|6|6|100.00%|

**Review by gengjun-git — APPROVED** [review]

**Pull request merged**